===================================================================
RCS file: /cvs/src/src/ld/emultempl/elf32.em,v
retrieving revision 1.131.2.1
retrieving revision 1.131.2.1.2.1
diff -u -r1.131.2.1 -r1.131.2.1.2.1
--- src/ld/emultempl/elf32.em	2005/03/22 14:49:54	1.131.2.1
+++ src/ld/emultempl/elf32.em	2005/04/13 14:58:17	1.131.2.1.2.1
@@ -226,6 +226,9 @@
     return;
   if (s->the_bfd == NULL)
     return;
+  if (s->as_needed
+      && (bfd_elf_get_dyn_lib_class (s->the_bfd) & DYN_AS_NEEDED) != 0)
+    return;
 
   if (bfd_stat (s->the_bfd, &st) != 0)
     {
@@ -733,6 +736,13 @@
   if (global_found)
     return;
 
+  /* If this input file was an as-needed entry, and wasn't found to be
+     needed at the stage it was linked, then don't say we have loaded it.  */
+  if (s->as_needed
+      && (s->the_bfd == NULL
+	  || (bfd_elf_get_dyn_lib_class (s->the_bfd) & DYN_AS_NEEDED) != 0))
+    return;
+
   if (s->filename != NULL)
     {
       const char *f;
